org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (c) 2012, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.xbase.ui.editor;
+
+import org.apache.log4j.Logger;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IStorage;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.ide.ResourceUtil;
+import org.eclipse.ui.part.FileEditorInput;
+import org.eclipse.xtext.LanguageInfo;
+import org.eclipse.xtext.resource.FileExtensionProvider;
+import org.eclipse.xtext.ui.editor.utils.EditorUtils;
+import org.eclipse.xtext.ui.generator.trace.IEclipseTrace;
+import org.eclipse.xtext.ui.generator.trace.ILocationInEclipseResource;
+import org.eclipse.xtext.ui.generator.trace.ITraceForStorageProvider;
+
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+
+public class XbaseEditorInputRedirector {
+
+	private final static Logger LOG = Logger.getLogger(XbaseEditorInputRedirector.class);
+
+	@Inject
+	private ITraceForStorageProvider traceInformation;
+
+	@Inject
+	private LanguageInfo languageInfo;
+
+	@Inject
+	private FileExtensionProvider fileExtensionProvider;
+
+	public ITypeRoot getTypeRoot(IEditorInput editorInput) {
+		IJavaElement adapter = editorInput.getAdapter(IJavaElement.class);
+		return adapter instanceof ITypeRoot ? (ITypeRoot) adapter : null;
+	}
+
+	/**
+	 * @return the original source for an editor input that points to an Xtext resource copied to the output folder, the given input
+	 *         otherwise
+	 */
+	public IEditorInput findOriginalSourceForOuputFolderCopy(IEditorInput input) {
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
```

Comment:
I would early return here as well, also this doesn't need to be in try catch, try to keep it as small as possible